File: engine/dynamic/render_vulkan/src/state/renderer_state.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/logging.hpp"

#include "argus/render/2d/scene_2d.hpp"
#include "argus/render/common/scene.hpp"

#include "internal/render_vulkan/state/renderer_state.hpp"
#include "internal/render_vulkan/state/scene_state.hpp"

namespace argus {
    SceneState &RendererState::get_scene_state(Scene &scene, bool create) {
        switch (scene.type) {
            case SceneType::TwoD: {
                auto &scene_2d = reinterpret_cast<const Scene2D &>(scene);
                auto it = this->scene_states_2d.find(&scene_2d);
                if (it != this->scene_states_2d.cend()) {
                    return it->second;
                }

                if (!create) {
                    Logger::default_logger().fatal("Failed to get scene state");
                }

                auto insert_res = this->scene_states_2d.try_emplace(&scene_2d, *this, scene);
                if (!insert_res.second) {
                    Logger::default_logger().fatal("Failed to create new scene state");
                }

                return insert_res.first->second;
            }
            case SceneType::ThreeD: {
                Logger::default_logger().fatal("Unimplemented scene type");
            }
            default: {
                Logger::default_logger().fatal("Unrecognized scene type");
            }
        }
    }

    ViewportState &RendererState::get_viewport_state(AttachedViewport &viewport, bool create) {
        switch (viewport.type) {
            case SceneType::TwoD: {
                auto &viewport_2d = reinterpret_cast<AttachedViewport2D &>(viewport);
                auto it = this->viewport_states_2d.find(&viewport_2d);
                if (it != this->viewport_states_2d.cend()) {
                    return it->second;
                }

                if (!create) {
                    Logger::default_logger().fatal("Failed to get viewport state");
                }

                Viewport2DState state(*this, &viewport_2d);
                auto insert_res = this->viewport_states_2d.insert({&viewport_2d, state});
                if (!insert_res.second) {
                    Logger::default_logger().fatal("Failed to create new viewport state");
                }

                return insert_res.first->second;
            }
            case SceneType::ThreeD: {
                Logger::default_logger().fatal("Unimplemented viewport type");
            }
            default: {
                Logger::default_logger().fatal("Unrecognized viewport type");
            }
        }
    }
}

File: engine/modules/render_opengl/src/state/renderer_state.rs
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::collections::HashMap;
use std::rc::Rc;
use argus_resman::{Resource, ResourceIdentifier};
use argus_wm::GlContext;
use crate::shaders::*;
use crate::state::*;
use crate::util::buffer::GlBuffer;
use crate::util::gl_util::*;

#[derive(Default)]
pub(crate) struct RendererState {
    pub gl_context: Option<GlContext>,
    pub intrinsic_resources: Vec<Resource>,
    pub scene_states_2d: HashMap<String, Scene2dState>,
    pub viewport_states_2d: HashMap<u32, ViewportState>,
    pub are_viewports_initialized: bool,
    pub prepared_textures: HashMap<String, Rc<GlTextureHandle>>,
    pub material_textures: HashMap<ResourceIdentifier, (String, Rc<GlTextureHandle>)>,
    pub compiled_shaders: HashMap<String, GlShaderHandle>,
    pub linked_programs: HashMap<ResourceIdentifier, LinkedProgram>,
    pub std_program: Option<LinkedProgram>,
    pub shadowmap_program: Option<LinkedProgram>,
    pub lighting_program: Option<LinkedProgram>,
    pub lightmap_composite_program: Option<LinkedProgram>,
    pub postfx_programs: HashMap<String, LinkedProgram>,
    pub frame_vbo: Option<GlBufferHandle>,
    pub frame_vao: Option<GlArrayHandle>,
    pub frame_program: Option<LinkedProgram>,
    pub global_ubo: Option<GlBuffer>,
}

impl<'a> RendererState {
    pub fn get_scene_2d_state(&'a self, scene_id: impl AsRef<str>) -> &'a Scene2dState {
        self.scene_states_2d.get(scene_id.as_ref()).expect("Scene state is missing")
    }

    pub fn get_scene_2d_state_mut(&'a mut self, scene_id: impl AsRef<str>) -> &'a mut Scene2dState {
        self.scene_states_2d.get_mut(scene_id.as_ref()).expect("Scene state is missing")
    }

    pub fn get_or_create_scene_2d_state(&'a mut self, scene_id: impl AsRef<str>)
        -> &'a mut Scene2dState {
        self.scene_states_2d.entry(scene_id.as_ref().to_string())
            .or_insert_with_key(|id| Scene2dState::new(id.clone()))
    }

    pub fn get_viewport_2d_state(&'a self, viewport_id: u32) -> &'a ViewportState {
        self.viewport_states_2d.get(&viewport_id)
            .expect("Viewport state is missing")
    }

    pub fn get_viewport_2d_state_mut(&'a mut self, viewport_id: u32) -> &'a mut ViewportState {
        self.viewport_states_2d.get_mut(&viewport_id)
            .expect("Viewport state is missing")
    }

    pub fn get_or_create_viewport_2d_state(&'a mut self, viewport_id: u32) -> &'a mut ViewportState {
        self.viewport_states_2d.entry(viewport_id)
            .or_insert_with(ViewportState::new)
    }
}
